### PR TITLE
Update 4x50

### PR DESCRIPTION
**Pull request by tharexde**
As discussed with iceman in advance, here's a nearly complete update on EM4x50. Sorry for not separating it in multiple pull requests. 
The following major points have been addressed:

Refactoring
- Removed global tag structure on device side
- Stabilized and accelerated tag detection
- Switched to cliparser for all functions
- Changed cli parameter: from w (word) to d (data)
- Changed cli parameter: from a (address) to b (block)
- All hex input parameters now have to be given in lsb format (output is still msb + lsb)
- Renamed 4x50_write_password to 4x50_writepwd
- Relocated write requests in function 4x50_wipe from device to client
- Added json format for 4x50_dump
- Added LED signals

New functions
- 4x50_eload: upload em4x50 file content (bin, eml, json) to flash memory
- 4x50_esave: dump em4x50 content from flash memory to file (bin + eml + json)
- 4x50_restore: restore dump file (bin, eml, json) onto tag (via flash memory)
- 4x50_sim: simulate dump from file (bin/eml/json) or flash
- 4x50_reader: read data from tag (configured data -> standard read mode) incl. option -@
- 4x50_chk: guess password from dictionary (without option -> T55xx default dictionary or -f user dictionary)
- 4x50_brute: guess password within a given password range
- 4x50_login: authenticate against tag
- 4x50_reset: reset tag

### CODE REVIEW

File: armsrc/Standalone/Makefile.hal [hunk 1]
```diff
@@ -70,7 +70,7 @@ endef
 STANDALONE_MODES := LF_SKELETON LF_EM4100EMUL LF_EM4100RSWB LF_EM4100RWC LF_HIDBRUTE LF_ICEHID LF_PROXBRUTE LF_SAMYRUN
 STANDALONE_MODES += HF_14ASNIFF HF_AVEFUL HF_BOG HF_COLIN HF_ICECLASS HF_LEGIC HF_MATTYRUN HF_MSDSAL HF_TCPRST HF_YOUNG
 STANDALONE_MODES_REQ_SMARTCARD :=
-STANDALONE_MODES_REQ_FLASH := LF_ICEHID HF_14ASNIFF HF_BOG HF_COLIN HF_ICECLASS
```
Comment (iceman1001) [inline armsrc/Standalone/Makefile.hal:73]: This one shouldn't be changed.

File: client/src/cmdlfem4x50.c [hunk 2]
```diff
@@ -246,270 +121,501 @@ static void print_info_result(uint8_t *data, bool verbose) {
...
-    int res = em4x50_read(&etd, words, false);
-    if (res == PM3_SUCCESS)
-        PrintAndLogEx(INFO, " Serial: " _GREEN_("%s"), sprint_hex(words[EM4X50_DEVICE_SERIAL].byte, 4));
-    return res;
+static int em4x50_wipe_flash(int page) {
+    
+    int isok = 0;
+    
+    clearCommandBuffer();
+    SendCommandMIX(CMD_FLASHMEM_WIPE, page, false, 0, NULL, 0);
```
Comment (iceman1001) [inline client/src/cmdlfem4x50.c:161]: flashmem_wipe??  ..
We use SPIFFS if you wanna save files.  
And this needs to be guarded with checks for RDV4 w flashmem enabled..

File: client/src/cmdlfem4x50.c [hunk 3]
```diff
@@ -246,270 +121,501 @@ static void print_info_result(uint8_t *data, bool verbose) {
...
-    if (errors)
-        return usage_lf_em4x50_info();
+    conn.block_after_ACK = false;
+    
+    return PM3_SUCCESS;
+}
 
-    clearCommandBuffer();
-    SendCommandNG(CMD_LF_EM4X50_INFO, (uint8_t *)&etd, sizeof(etd));
+int CmdEM4x50ELoad(const char *Cmd) {
```
Comment (iceman1001) [inline client/src/cmdlfem4x50.c:222]: eload in the client is used for "emulator memory".  A specific part of Bigbuf ,  ram memory.
This seems to be a flash memory now,   which is specific for RDV4.  ... ...

File: client/src/cmdlfem4x50.c [hunk 4]
```diff
@@ -246,270 +121,501 @@ static void print_info_result(uint8_t *data, bool verbose) {
...
+    };
+
+    CLIExecWithReturn(ctx, Cmd, argtable, true);
+    
+    CLIGetHexWithReturn(ctx, 1, pwd, &pwdLen);
+    if (pwdLen != 4) {
+        PrintAndLogEx(FAILED, "password length must be 4 bytes instead of %d", pwdLen);
+        return PM3_EINVARG;
+    } else {
+        password = (pwd[0] << 24) | (pwd[1] << 16) | (pwd[2] << 8) | pwd[3];
```
Comment (iceman1001) [inline client/src/cmdlfem4x50.c:350]: BSWAP32 ?

File: client/src/cmdlfem4x50.c [hunk 5]
```diff
@@ -246,270 +121,501 @@ static void print_info_result(uint8_t *data, bool verbose) {
...
-                    PrintAndLogEx(FAILED, "\n  password has to be 8 hex symbols\n");
-                    return PM3_EINVARG;
-                }
-                bpwd = true;
-                etd.pwd_given = true;
-                cmdp += 2;
-                break;
+    CLIParserContext *ctx;
+    CLIParserInit(&ctx, "lf em 4x50_reset",
+                  "Reseta EM4x50 tag.",
```
Comment (iceman1001) [inline client/src/cmdlfem4x50.c:561]: Maybe some more text explaining what a em4x50 reset is and does

File: client/src/cmdlfem4x50.c [hunk 6]
```diff
@@ -527,225 +633,684 @@ int em4x50_read(em4x50_data_t *etd, em4x50_word_t *out, bool verbose) {
...
+    CLIParserContext *ctx;
+    CLIParserInit(&ctx, "lf em 4x50_read",
+                  "Reads single EM4x50 block/word.",
+                  "lf em 4x50_read -b 3 -> reads block 3\n"
+                  "lf em 4x50_read -b 32 -p 12345678 -> reads block 32 with password 0x12345678\n"
+                 );
+
+    void *argtable[] = {
+        arg_param_begin,
+        arg_int1("b", "block", "<address>", "block/word address, dec"),
```
Comment (iceman1001) [inline client/src/cmdlfem4x50.c:672]: <address>  ->   <dec>

File: client/src/cmdlfem4x50.c [hunk 7]
```diff
@@ -527,225 +633,684 @@ int em4x50_read(em4x50_data_t *etd, em4x50_word_t *out, bool verbose) {
...
+        arg_str0("p", "passsword", "<password>", "password, hex, 4 bytes, lsb"),
```
Comment (iceman1001) [inline client/src/cmdlfem4x50.c:673]: <password>   ->  <hex>

File: client/src/cmdlfem4x50.c [hunk 8]
```diff
@@ -527,225 +633,684 @@ int em4x50_read(em4x50_data_t *etd, em4x50_word_t *out, bool verbose) {
...
+                     );
+    CLIGetHexWithReturn(ctx, 3, pwd, &pwdLen);
+    
+    if ((uidLen && fnLen) || (!uidLen && !fnLen)) {
+        PrintAndLogEx(FAILED, "either use option 'u' or option 'f'");
+        return PM3_EINVARG;
+    }
+    
+    if (uidLen) {
+        snprintf(filename, FILE_PATH_SIZE, "./lf-4x50-%02x%02x%02x%02x-dump.bin",
```
Comment (iceman1001) [inline client/src/cmdlfem4x50.c:1195]: I think we have a fct "fillbyuid"  to generate a filename based on uid and template.
I am not sure about your  "./"   including of the file name.

**Comment by iceman1001** [comment]
Yeah,  like we talked about its a massive PR.  
As I see it now,  the confusion about "eload" / "eset" compared with how its used with the similar e* commands in the client is the largest issue.

**Comment by tharexde** [comment]
My understanding from a previous discussion was that we talked explicitly about flash. Actually I started with emulator memory and then moved on to flash. Ok, then I'll go back to the emulator version for eload and esave.

**Comment by iceman1001** [comment]
Look into the SPIFFS commands in order to use files uploaded.   However you will load from SPIFFS in to the emulator memory,   which is used as tag memory for the simulation.  Its a bit confusing.  The biggest issue is that only RDV4 has SPIFFS,   and we wanna make sure much devices can use em4x50 sim also.

**Comment by tharexde** [comment]
Ok, I'll check and correct your above mentioned points (and probably there are one or two more...).

**Review by BigD1CKC — APPROVED** [review]

**Review by BigD1CKC — APPROVED** [review]

File: client/src/cmdlfem4x50.c (hunk 7 above)
Comment (tharexde) [reply client/src/cmdlfem4x50.c:673]: Maybe a hint what you mean?

Comment (tcprst) [reply client/src/cmdlfem4x50.c:673]: I think github filtered out what looked like HTML tags from his comment, somehow they came through in the Discord integration.  `<password> -> <hex>` is what it said

File: client/src/cmdlfem4x50.c (hunk 6 above)
Comment (tcprst) [reply client/src/cmdlfem4x50.c:672]: This one said `<address> -> <dec>`

File: client/src/cmdlfem4x50.c (hunk 7 above)
Comment (tharexde) [reply client/src/cmdlfem4x50.c:673]: Ok, thanks.

File: client/src/cmdlfem4x50.c (hunk 8 above)
Comment (tharexde) [reply client/src/cmdlfem4x50.c]: fixed

File: armsrc/Standalone/Makefile.hal (hunk 1 above)
Comment (tharexde) [reply armsrc/Standalone/Makefile.hal]: fixed

File: client/src/cmdlfem4x50.c (hunk 6 above)
Comment (iceman1001) [reply client/src/cmdlfem4x50.c:672]: Seems GH didn't like the arrows

**Comment by doegox** [comment]
Thanks a lot for your contributions!
Generally speaking, it's much easier to deal with smaller PRs than such massive one, including in the event a bug much be chased with git bisect.
The fact that you mentioned in the changelog 19 distinct lines is a clear indication this PR could have been split into several PRs.

**Comment by iceman1001** [comment]
Yes,  I had a conversation with @tharexde  before he made this PR where we agreed this will be the best way forward.
I suppose @tharexde  that your standalone mode can be excluded in this PR ,  to make it a bit smaller.

**Comment by iceman1001** [comment]
Appveyor complains on this PR,
image

**Comment by iceman1001** [comment]
Excellent,
I merge it and lets see what happens,  since it is a large PR.

**Pull request merged**